src/audio/eq_iir/eq_iir.c

```diff
@@ -792,16 +828,24 @@ static int eq_iir_prepare(struct processing_module *mod)
 
 	comp_dbg(dev, "eq_iir_prepare()");
 
+	/* TODO: This is untested, may not work in prepare() */
```

Comment:
Hmm, this TODO raises more questions than it answers. We use this code path every time, so isn't it problematic if is not tested...?

Reply:
I should have been more specific. Setting different word length for source and sink may not work with module adapter since it's hiding params() currently from clients. The purpose of this params function is the check if IIR has conversion for source -> sink format. If yes, keep it. Otherwise set sink format to source. @ranj063 Do you have thoughts for this?

Reply:
By the the time at prepare() I don't think there can be any more different source and sink format. So this may be useless here.

Reply:
@singalsu this line here in verify_params seems wrong:
https://github.com/thesofproject/sof/blob/478a2bee5c880aafa3d1aab01e7812e202d78357/src/audio/eq_iir/eq_iir.c#L690

the eq_iir_find_func() returns a function pointer but buffer_flag is an uint32_t. Im not sure what the original intention was for the buffer_flag to be but seems pretty useless to me.

Reply:
Yep, that seems to be totally broken as well. Behavior is random depending on function address bits that matches flag definitions. The intention might have been to set BUFF_PARAMS_BUFFER_FMT if address is not NULL.
I feel I should remove this call and the eq_iir_verify_params() function. It will remove the format conversion capability from IIR. SOF v2.2 might need a fix. IPC3 will work with format conversion in host/DAI. In IPC4 topologies the pipelines seem to be fixed 32 bit format and all conversion happens in host/DAI copiers.

Do you agree @ranj063 and @kv2019i ?

Reply:
Oops, I missed that it's `buffer_flag = eq_iir_find_func() ? BUFF_PARAMS_FRAME_FMT : 0;` I'll check with IPC3 topology format conversion in IIR before and after module adapter to help decide what to do.

Reply:
I just checked with git main that with IPC3 and sof-hda-generic-4ch.tplg the 48 kHz capture with `arecord -Dhw:0,6 -f S16_LE -c 4 -r 48000 -d 10 rec.wav` the IIR converts the format from 32 to 16 bit. It gives the best 16 bit quality from 32 bit DAI. The trace evidence is "eq_iir_prepare(), source_format=2, sink_format=0". With this module adapter version of PR with eq_iir_verify_params() called from prepare() the trace is "eq_iir_prepare(), source_format=0, sink_format=0" that means that DAI calls PCM converter and IIR is all 16 bits. The parameters verification in prepare() does no more have  opportunity to volunteer for format conversion.

@lgirdwood @ranj063 @kv2019i It's clear that enabling format conversion from prepare does not work. Should we add to module adapter a params() hook for components those can change stream format with IPC3? Or Drop this feature for now from IPC3 support of git main and see later how this could be re-introduced with IPC4 if needed?

Reply:
@singalsu I think this could be kept as well. With current IPC4 topologies (and kernel code), the source format is fixed, so the conversion capability is neverused (IIR will always see S32 from the DAI). I currently have a kernel PR in review to change specifically this, so you could define a topology, where DAI provides other formats.  So then at prepare you'd see  also different source formats.

But the other question is whether there is need to have this code in other components. In IPC4, we have a copier at the DAI and the FE PCM, and they both can convert. So I guess the main benefit would be save a bit of compute by pushing the conversion from DAI copier to IIR output. But not sure this is worth the additional code (to maintain and have in the module).

Reply:
Thanks! OK, I can keep it there since seems to be no harm. Though the conversion feature will break with module_adapter since the params() walk handling is not complete.